`DRAFT` 🗃 Feature for Compare Databases

edelciomolina opened the pull request:
Includes new functionality that compares functions, procedures, tables, indexes, etc., returning a detailed object for preparatory analysis of future dump or restore.

![image](https://github.com/user-attachments/assets/ecb363ed-dd5a-4a68-966b-4d6b5f7015d3)

edelciomolina committed:
Add Verbose Argument

edelciomolina committed:
Add: Workaround for Create argument

edelciomolina committed:
Add PG package

edelciomolina committed:
Add createMethod and createPsqlWith

edelciomolina committed:
fix: throw error

edelciomolina committed:
fix: throw error

edelciomolina committed:
fix: throw error

edelciomolina committed:
Fix: Log Error createDatabaseAndRetry

edelciomolina committed:
Fix: Check pg_restore: error:

edelciomolina committed:
Fix already exists

edelciomolina committed:
Fix already exists

edelciomolina committed:
Fix already exists

edelciomolina committed:
Add Improvements after production running

edelciomolina committed:
Add: Compare Method

jdexyz commented:
This looks like a very detailed comparison indeed.
- Do you think it should be shipped with this package which is focused on pg-restore and pg-dump, rather than being its own package ? 
- I think such a complex feature really needs an automated test 🙂

edelciomolina commented:
> This looks like a very detailed comparison indeed.
> 
> * Do you think it should be shipped with this package which is focused on pg-restore and pg-dump, rather than being its own package ?
> * I think such a complex feature really needs an automated test 🙂

In my opinion, I think it's an important method to complement your wrapper. In my case, I'm working on an application to transfer databases from AWS to a local VM, and to avoid redoing the dump/copy work, I thought it would be prudent to have a functionality, at least a basic one, to indicate that if there are still differences, the process should continue.

But I think that due to a lack of care, I ended up mixing up PRs, and this one should be based on the MASTER. So, let me work on a new PR with this implementation, and I will also include some tests.

edelciomolina committed:
Add: checkError

Ensure that in case it is really an error, then use console.error. Avoiding the output as an error in cases of warning that comes via stderr

edelciomolina committed:
Add: Win Binary Versions

Add v9.3.25, v15.3, v16.3, v16.4 and the lastest 17rc1 as default.

edelciomolina committed:
Add: Check Dump With Verbose

edelciomolina committed:
Merge branch 'master' of https://github.com/EndoData/pg-dump-restore-node-wrapper into pg-versions-bin

# Conflicts:
#	index.js

edelciomolina committed:
Add: Tests for dump and restore with verbose args

edelciomolina committed:
Rebase: From master

edelciomolina committed:
Fix checkError in Dump

edelciomolina committed:
Add caniuse-lite for Jest

edelciomolina committed:
Fix: Creating database (if requested with psql)

edelciomolina committed:
Refactor: Tests

Creating and droping database.

edelciomolina committed:
Add: Extra Mock for testing

edelciomolina committed:
Add: New tests

edelciomolina committed:
Avoid pgdump files to repo

edelciomolina committed:
Change: Create and Clean Methods

Add some important comments

edelciomolina committed:
Fix: Documentation

edelciomolina committed:
Merge branch 'master' of https://github.com/EndoData/pg-dump-restore-node-wrapper into create-and-retry

# Conflicts:
#	index.js

edelciomolina committed:
Fix: Last Merge without semi

edelciomolina committed:
Fix: Description Tests

edelciomolina committed:
Fix: Description Tests

edelciomolina committed:
Merge branch 'add-verbose' of https://github.com/edelciomolina/pg-dump-restore-node-wrapper into create-and-retry

# Conflicts:
#	index.js
#	tests/main.test.js

edelciomolina committed:
Add: Extra check if dump file exists before restore

edelciomolina committed:
Fix: Verbose at cleanDatabase

edelciomolina committed:
Improvement: Normalizing PSQL functions into separate Lib.

This was necessary because index.js was using mock functions.

edelciomolina committed:
Rollback: Bin 13.4

edelciomolina committed:
Fix: Readme with version args

edelciomolina committed:
Merge branch 'master' of https://github.com/EndoData/pg-dump-restore-node-wrapper into feat-compare

# Conflicts:
#	index.js
#	readme.md

edelciomolina committed:
Merge branch 'create-and-retry' of https://github.com/edelciomolina/pg-dump-restore-node-wrapper into feat-compare

# Conflicts:
#	index.js
#	package-lock.json
#	readme.md

edelciomolina committed:
Merge branch 'pg-versions-bin' of https://github.com/edelciomolina/pg-dump-restore-node-wrapper into feat-compare

# Conflicts:
#	index.js
#	package-lock.json

edelciomolina committed:
Add: Lib Folders

edelciomolina commented:
- [x] Pulled https://github.com/EndoData/pg-dump-restore-node-wrapper/pull/4
- [x] Pulled https://github.com/EndoData/pg-dump-restore-node-wrapper/pull/5
- [x] Pulled https://github.com/EndoData/pg-dump-restore-node-wrapper/pull/8
- [ ] Create tests for feature Compare Lib
  - [ ] Row count mismatch
  - [ ] Has N mismatch rows in source
  - [ ] Has N mismatch rows in target
  - [ ] Table exists in both source and target with matching rows
  - [ ] Table exists in source but not in target
  - [ ] Table exists in target but not in source
  - [ ] Function exists in both source and target
  - [ ] Function exists in source but not in target
  - [ ] Function exists in target but not in source
  - [ ] Index exists in both source and target
  - [ ] Index exists in source but not in target
  - [ ] Index exists in target but not in source
  - [ ] Procedure exists in both source and target
  - [ ] Procedure exists in source but not in target
  - [ ] Procedure exists in target but not in source

edelciomolina committed:
Fix: Tests

Now it's passing.

edelciomolina committed:
Fix: Moving compare.js to libs and fix them

edelciomolina committed:
Fix: Avoid database locks in parallel testing

edelciomolina committed:
Fix: Making each test individual without depending on beforeAll or dependency on a previous test mock.

edelciomolina committed:
Add: Example (WIP)

edelciomolina committed:
Add: Example

edelciomolina committed:
Remove: Compare

Now all is update and ready to merge.

edelciomolina committed:
Merge pull request #1 from edelciomolina/refactor-without-compare

Refactor

edelciomolina committed:
Remove: Compare Docs

edelciomolina committed:
Merge pull request #2 from edelciomolina/refactor-without-compare

Remove: Compare Docs

edelciomolina committed:
Merge branch 'master' of https://github.com/edelciomolina/pg-dump-restore-node-wrapper into feat-compare

edelciomolina committed:
Add: Compare v2
